### PR TITLE
PLU-261: [onboarding-demo-experiment-3]: add basic demo flow to flows page

### PR DESCRIPTION
**Pull request by m0nggh**
## Problem
New users are confused with how to get started on Plumber, so GGWP v1 is an onboarding demo flow we are experimenting with to provide more help for users to onboard with Plumber.

## Details
Frontend: 
- Demo flow appears on the flows page, and it allow users to click in, `onFirstload` being true should render 2 things
    1. The tooltip on hover
    2. The `DemoFlowModal`

General flow
- User clicks on the first demo pipe --> it leads to the `DemoFlowModal` inside the flow editor
  - Click on build your own workflow or anywhere outside allows user to immediately work on the flow
  - Click on start demo leads to the demo video, click anywhere outside also allows user to work on the flow

Backend: 
- Update the flow config upon onClose (inside the editor already)
- Duplicate pipe currently duplicates the config —> 
- [x] **TODO:  remove `duplicateCount` and `demoConfig`

## Tests
- [x] Can view demo page modal and the video inside the flow editor
- [x] Non-demo pipes will load and still work as intended
- [x] Ensure that context menu view button also has the same functionality: demo opens modal but normal pipes load into the editor

### CODE REVIEW

**Comment by m0nggh** [comment]
* **#622** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/622?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#621** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/621?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#620** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/620?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#619** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/619?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#618** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/618?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#617** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/617?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop-v2`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @m0nggh and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/backend/src/graphql/mutations/duplicate-flow.ts [hunk 1]
```diff
@@ -75,21 +75,28 @@ const duplicateFlow: MutationResolvers['duplicateFlow'] = async (
...
     })
 
+    // duplicate the flow with the previous config
     const duplicatedFlow = await context.currentUser
       .$relatedQuery('flows', trx)
       .insert({
         name: `[COPY] ${flow.name}`,
         active: false,
+        config:
+          Object.keys(prevConfig).length > 0 ? { ...prevConfig } : undefined,
```
Comment (pregnantboy) [inline packages/backend/src/graphql/mutations/duplicate-flow.ts]: nit: _.isEmtpy and no need to re-spread prevConfig

File: packages/backend/src/helpers/csp.ts [hunk 2]
```diff
@@ -22,6 +22,7 @@ const helmetOptions: HelmetOptions = {
       frameAncestors: ["'none'"],
       frameSrc: [
         "'self'",
+        'https://demo.arcade.software',
```
Comment (pregnantboy) [inline packages/backend/src/helpers/csp.ts:25]: [goodcatch](https://app.graphite.dev//settings/meme-library?org=opengovsg)

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 3]
```diff
@@ -40,11 +43,43 @@ export default function EditorLayout(): React.ReactElement {
...
+
+  const [updateFlowConfig] = useMutation(UPDATE_FLOW_CONFIG)
+  const onFlowConfigUpdate = useCallback(async () => {
+    await updateFlowConfig({
+      variables: {
+        input: {
+          id: flow?.id,
+          onFirstLoad: false, // this is to not load demo modal and show tooltip anymore
+        },
+        optimisticResponse: {
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx]: unnecessary

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 4]
```diff
@@ -40,11 +43,43 @@ export default function EditorLayout(): React.ReactElement {
   })
   const flow: IFlow = data?.getFlow
 
+  // for loading demo modal
+  const onFirstLoad = flow?.config?.demoConfig?.onFirstLoad
+  const { onClose } = useDisclosure()
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx]: no-op can remove, ure not using `isOpen` at all

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 5]
```diff
@@ -40,11 +43,43 @@ export default function EditorLayout(): React.ReactElement {
...
+              },
+            },
+          },
+        },
+      },
+    })
+  }, [flow?.id, updateFlowConfig])
+
+  const handleClose = useCallback(() => {
+    onClose()
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx:74]: can remove

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 6]
```diff
@@ -189,6 +224,8 @@ export default function EditorLayout(): React.ReactElement {
         isOpen={!!flow?.active}
         handleUnpublish={() => onFlowStatusUpdate(!flow.active)}
       ></EditorSnackbar>
+
+      {onFirstLoad && <DemoFlowModal onClose={handleClose} />}
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx:228]: can call updateFlowFirstLoad directly

File: packages/frontend/src/components/FlowRow/DemoFlowModal.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      motionPreset="none"
+      closeOnEsc={false}
+    >
+      <ModalOverlay bg="base.canvas.overlay" />
+      <ModalContent my={12}>
+        {showVideoModal ? (
+          // TODO (mal): to update this when there are more demo links in the future
+          <DemoVideoModalContent
+            src="https://demo.arcade.software/FzpL1zCmibw0oXR6HUJi?embed&show_copy_link=true"
+            title={`Forms -&gt; Postman (GGWP)`}
```
Comment (pregnantboy) [inline packages/frontend/src/components/FlowRow/DemoFlowModal.tsx:39]: remove unnecessary

File: packages/frontend/src/components/EditorLayout/index.tsx [hunk 8]
```diff
@@ -40,11 +43,43 @@ export default function EditorLayout(): React.ReactElement {
   })
   const flow: IFlow = data?.getFlow
 
+  // for loading demo modal
+  const onFirstLoad = flow?.config?.demoConfig?.onFirstLoad
+  const { onClose } = useDisclosure()
+
+  const [updateFlowConfig] = useMutation(UPDATE_FLOW_CONFIG)
+  const onFlowConfigUpdate = useCallback(async () => {
```
Comment (pregnantboy) [inline packages/frontend/src/components/EditorLayout/index.tsx]: can just add the variables inside useMutation, no ned to create another callback

File: packages/frontend/src/components/EditorLayout/index.tsx (hunk 4 above)
Comment (pregnantboy) [reply packages/frontend/src/components/EditorLayout/index.tsx]: actually can get use to closeModal immediately.
```
const {onOpen, onClose, isOpen} = useDisclosure()
useEffect(() => {
   if (onFirstLoad) { 
     onOpen() 
   }
}, [onFirstLoad])
```

File: packages/frontend/src/components/FlowRow/DemoFlowModal.tsx (hunk 7 above)
Comment (m0nggh) [reply packages/frontend/src/components/FlowRow/DemoFlowModal.tsx:39]: actly can we just leave the title for `iframe` for accessibility purposes?

edit: anyway i will move this into a frontend file to store all the templates in another PR too

File: packages/frontend/src/components/EditorLayout/index.tsx (hunk 4 above)
Comment (m0nggh) [reply packages/frontend/src/components/EditorLayout/index.tsx]: hmm actually i think i just don't use it baa thks

File: packages/backend/src/graphql/mutations/duplicate-flow.ts (hunk 1 above)
Comment (m0nggh) [reply packages/backend/src/graphql/mutations/duplicate-flow.ts]: i only keep the notification frequency config, the rest will be removed (demo config and duplicate count)

**Review by pregnantboy — APPROVED** [review]

**Pull request merged**